src/sagemaker_training/pytorch_xla.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2018-2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License'). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the 'license' file accompanying this file. This file is
+# distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""This module contains functionality related to distributed training using
+PT-XLA (PyTorch - Accelerated Linear Algebra)."""
+from __future__ import absolute_import
+
+import os
+
+from sagemaker_training import (
+    _entry_point_type,
+    environment,
+    errors,
+    logging_config,
+    process,
+)
+
+
+logger = logging_config.get_logger()
+
+
+class PyTorchXLARunner(process.ProcessRunner):
+    """Responsible for PT-XLA distributed training."""
+
+    MESH_SERVICE_PORT = 53957
+    WORKER_PORT = 43857
+
+    def __init__(
+        self,
+        user_entry_point,
+        args,
+        env_vars,
+        processes_per_host,
+        master_hostname,
+        current_host,
+        hosts,
+        num_gpus,
+    ):
+        """Initialize a PyTorchXLARunner, which is responsible for distributed
+        training with PT-XLA.
+
+        Args:
+            user_entry_point (str): The name of the user entry point.
+            args ([str]): A list of arguments to include when executing the entry point.
+            env_vars (dict(str,str)): A dictionary of environment variables.
+            master_hostname (str): The master hostname.
+            current_host (str): The current hostname.
+            hosts ([str]): A list of hosts.
+            num_gpus (int): The number of GPUs available per host.
+        """
+
+        super(PyTorchXLARunner, self).__init__(user_entry_point, args, env_vars, processes_per_host)
+
+        self._master_hostname = master_hostname
+        self._current_host = current_host
+        self._hosts = hosts
+        self._num_gpus = num_gpus
+
+        self._num_hosts = len(self._hosts)
+        self._rank = self._hosts.index(self._current_host)
+
+    def _setup(self):  # type: () -> None
+        logger.info("Starting distributed training through PT-XLA Runtime.")
+        self._check_compatibility()
+
+        os.environ["XRT_HOST_ORDINAL"] = str(self._rank)
+        os.environ["XRT_SHARD_WORLD_SIZE"] = str(self._num_hosts)
+        address = "localservice:{};{}:" + str(self.WORKER_PORT)
+        os.environ["XRT_WORKERS"] = "|".join(
+            [address.format(i, host) for i, host in enumerate(self._hosts)]
+        )
+        os.environ["GPU_NUM_DEVICES"] = str(self._num_gpus)
```

Comment:
We need to set some extra environment variables to ensure EFA works correctly with NCCL. Please add the following:

```
# Add configuration for NCCL & EFA NCCL Plugin

# Set NCCL logging to info to debug
# customers issues easier
os.environ["NCCL_DEBUG"]="info"

# Use `simple` protocol to handle
# the out of order data delivery from EFA
os.environ["NCCL_PROTO"]="simple"

# use GPU RDMA when available
# applicable only to p4d.24xlarge
os.environ["FI_EFA_USE_DEVICE_RDMA"]="1"

# Use multiple connections per GPU to
# better saturate the EFA bandwith
os.environ["OFI_NCCL_NIC_DUP_CONNS"]=str(args.num_gpus)
```